File: packages/perspective/bench/js/perspective.benchmark.js
/******************************************************************************
 *
 * Copyright (c) 2017, the Perspective Authors.
 *
 * This file is part of the Perspective library, distributed under the terms of
 * the Apache License 2.0.  The full license can be found in the LICENSE file.
 *
 */

const CSV = "https://unpkg.com/@jpmorganchase/perspective-examples@0.2.0-beta.2/build/superstore.csv";
const ARROW = "https://unpkg.com/@jpmorganchase/perspective-examples@0.2.0-beta.2/build/superstore.arrow";

// Run entire test suite

const AGG_OPTIONS = [[{column: "Sales", op: "sum"}], [{column: "State", op: "dominant"}], [{column: "Order Date", op: "dominant"}]];
const COLUMN_PIVOT_OPTIONS = [[], ["Sub-Category"]];
const ROW_PIVOT_OPTIONS = [[], ["State"], ["State", "City"]];
const COLUMN_TYPES = {Sales: "number", "Order Date": "datetime", State: "string"};

const wait_for_perspective = () => new Promise(resolve => window.addEventListener("perspective-ready", resolve));

function to_name({aggregate, row_pivot, column_pivot}) {
    return `${COLUMN_TYPES[aggregate[0].column]},${row_pivot.join("/") || "-"}*${column_pivot.join("/") || "-"}`;
}

async function get_data(worker) {
    const req1 = fetch(CSV);
    const req2 = fetch(ARROW);

    console.log("Downloading CSV");
    let content = await req1;
    const csv = await content.text();

    console.log("Downloading Arrow");
    content = await req2;
    const arrow = await content.arrayBuffer();

    console.log("Generating JSON");
    const tbl = worker.table(arrow.slice());
    const view = tbl.view();
    const rows = await view.to_json();
    const columns = await view.to_columns();
    view.delete();
    tbl.delete();

    return {csv, arrow, rows, columns};
}

// eslint-disable-next-line no-unused-vars
async function run_test() {
    // eslint-disable-next-line no-undef
    perspective = perspective.default || perspective;

    const suite = new Suite("perspective");
    let worker, data;

    suite.beforeAll(async () => {
        worker = window.perspective.worker();
        await wait_for_perspective();
        data = await get_data(worker);
    });

    suite.description("table()", async () => {
        let table;

        suite.afterEach(async () => {
            await table.delete();
        });

        for (const name of Object.keys(data)) {
            suite.benchmark(name, async () => {
                let test = data[name];
                table = worker.table(test.slice ? test.slice() : test);
                await table.size();
            });
        }
    });

    suite.description("view()", async () => {
        let table;

        suite.beforeAll(async () => {
            table = worker.table(data.arrow.slice());
        });

        suite.afterAll(async () => {
            await table.delete();
        });

        for (const aggregate of AGG_OPTIONS) {
            for (const row_pivot of ROW_PIVOT_OPTIONS) {
                for (const column_pivot of COLUMN_PIVOT_OPTIONS) {
                    const config = {aggregate, row_pivot, column_pivot};

                    suite.description(`view(${to_name(config)}))`, async () => {
                        let view;

                        suite.afterEach(async () => {
                            await view.delete();
                        });

                        suite.benchmark(`view()`, async () => {
                            view = table.view(config);
                            await view.schema();
                        });
                    });

                    suite.description(`to_format(${to_name(config)}))`, async () => {
                        let view;

                        suite.beforeAll(async () => {
                            view = table.view(config);
                            await view.schema();
                        });

                        suite.afterAll(async () => {
                            await view.delete();
                        });

                        for (const format of ["to_json", "to_columns", "to_arrow"]) {
                            suite.benchmark(`${format}()`, async () => {
                                await view[format]();
                            });
                        }
                    });
                }
            }
        }
    });

    return await suite.run();
}

File: python/perspective/bench/perspective_benchmark.py
################################################################################
#
# Copyright (c) 2019, the Perspective Authors.
#
# This file is part of the Perspective library, distributed under the terms of
# the Apache License 2.0.  The full license can be found in the LICENSE file.
#

import os
import sys
from functools import partial
from bench import Benchmark, Suite, Runner
sys.path.insert(1, os.path.join(os.path.dirname(__file__), '..'))
from perspective import Table  # noqa: E402
from perspective.tests.common import superstore  # noqa: E402

SUPERSTORE = superstore(9994)

SUPERSTORE_ARROW = os.path.join(
    os.path.dirname(__file__),
    "..", "..", "..",
    "examples",
    "simple",
    "superstore.arrow")


def make_meta(group, name):
    return {
        "group": group,
        "name": name,
        "version": "master"
    }


class PerspectiveBenchmark(Suite):

    AGG_OPTIONS = [
        [{"column": "Sales", "op": "sum"}],
        [{"column": "State", "op": "dominant"}],
        [{"column": "Order Date", "op": "dominant"}]
    ]
    COLUMN_PIVOT_OPTIONS = [[], ["Sub-Category"], ["Category", "Sub-Category"]]
    ROW_PIVOT_OPTIONS = [[], ["State"], ["State", "City"]]

    VERSION = "master"

    def __init__(self):
        """Create a benchmark suite for `perspective-python`."""
        tbl = Table(SUPERSTORE)
        self._view = tbl.view()
        self.dict = self._view.to_dict()
        self.records = self._view.to_records()
        self.df = SUPERSTORE
        self.csv = self._view.to_csv()
        self.arrow = self._view.to_arrow()
        self._table = tbl

    def register_benchmarks(self):
        """Register all the benchmark methods - each method creates a number of
        lambdas, and then calls `setattr` on the Suite itself so that the
        `Runner` can find the tests at runtime."""
        self.benchmark_table()
        self.benchmark_table_arrow()
        self.benchmark_view_zero()
        self.benchmark_view_one()
        self.benchmark_view_two()
        self.benchmark_view_two_column_only()
        self.benchmark_to_format_zero()
        self.benchmark_to_format_one()
        self.benchmark_to_format_two()
        self.benchmark_to_format_two_column_only()

    def benchmark_table(self):
        """Benchmark table creation from different formats."""
        for name in ("df", "dict", "records"):
            data = getattr(self, name)
            test_meta = make_meta("table", name)
            func = Benchmark(lambda: Table(data), meta=test_meta)
            setattr(self, "table_{0}".format(name), func)

    def benchmark_table_arrow(self):
        """Benchmark table from arrow separately as it requires opening the
        Arrow file from the filesystem."""
        with open(SUPERSTORE_ARROW, "rb") as arrow:
            data = arrow.read()
            test_meta = make_meta("table", "arrow")
            func = Benchmark(lambda: Table(data), meta=test_meta)
            setattr(self, "table_arrow", func)

    def benchmark_view_zero(self):
        """Benchmark view creation with zero pivots."""
        func = Benchmark(lambda: self._table.view(), meta=make_meta("view", "zero"))
        setattr(self, "view_zero", func)

    def benchmark_view_one(self):
        """Benchmark view creation with different pivots."""
        for pivot in PerspectiveBenchmark.ROW_PIVOT_OPTIONS:
            if len(pivot) == 0:
                continue
            test_meta = make_meta("view", "one_{0}_pivot".format(len(pivot)))
            view_constructor = partial(self._table.view, row_pivots=pivot)
            func = Benchmark(lambda: view_constructor(), meta=test_meta)
            setattr(self, "view_{0}".format(test_meta["name"]), func)

    def benchmark_view_two(self):
        """Benchmark view creation with row and column pivots."""
        for i in range(len(PerspectiveBenchmark.ROW_PIVOT_OPTIONS)):
            RP = PerspectiveBenchmark.ROW_PIVOT_OPTIONS[i]
            CP = PerspectiveBenchmark.COLUMN_PIVOT_OPTIONS[i]
            if len(RP) == 0 and len(CP) == 0:
                continue
            test_meta = make_meta(
                "view", "two_{0}x{1}_pivot".format(len(RP), len(CP)))
            view_constructor = partial(
                self._table.view, row_pivots=RP, column_pivots=CP)
            func = Benchmark(lambda: view_constructor(), meta=test_meta)
            setattr(self, "view_{0}".format(test_meta["name"]), func)

    def benchmark_view_two_column_only(self):
        """Benchmark column-only view creation."""
        for pivot in PerspectiveBenchmark.COLUMN_PIVOT_OPTIONS:
            if len(pivot) == 0:
                continue
            test_meta = make_meta(
                "view", "two_column_only_{0}_pivot".format(len(pivot)))
            view_constructor = partial(self._table.view, column_pivots=pivot)
            func = Benchmark(lambda: view_constructor(), meta=test_meta)
            setattr(self, "view_{0}".format(test_meta["name"]), func)

    def benchmark_to_format_zero(self):
        """Benchmark each `to_format` method."""
        for name in ("numpy", "dict", "records", "df", "arrow"):
            test_meta = make_meta("to_format", name)
            func = Benchmark(
                lambda: getattr(self._view, "to_{0}".format(name))(), meta=test_meta)
            setattr(self, "to_format_{0}".format(name), func)

    def benchmark_to_format_one(self):
        """Benchmark each `to_format` method for one-sided contexts."""
        for name in ("numpy", "dict", "records", "df", "arrow"):
            for pivot in PerspectiveBenchmark.ROW_PIVOT_OPTIONS:
                if len(pivot) == 0:
                    continue
                test_meta = make_meta(
                    "to_format", "{0}_{1}".format(name, len(pivot)))
                view = self._table.view(row_pivots=pivot)
                func = Benchmark(
                    lambda: getattr(view, "to_{0}".format(name))(), meta=test_meta)
                setattr(self, "to_format_{0}".format(test_meta["name"]), func)

    def benchmark_to_format_two(self):
        """Benchmark each `to_format` method for two-sided contexts."""
        for name in ("numpy", "dict", "records", "df", "arrow"):
            for i in range(len(PerspectiveBenchmark.ROW_PIVOT_OPTIONS)):
                RP = PerspectiveBenchmark.ROW_PIVOT_OPTIONS[i]
                CP = PerspectiveBenchmark.COLUMN_PIVOT_OPTIONS[i]
                if len(RP) == 0 and len(CP) == 0:
                    continue
                test_meta = make_meta(
                    "to_format", "{0}_{1}x{2}".format(name, len(RP), len(CP)))
                view = self._table.view(row_pivots=RP, column_pivots=CP)
                func = Benchmark(
                    lambda: getattr(view, "to_{0}".format(name))(), meta=test_meta)
                setattr(self, "to_format_{0}".format(test_meta["name"]), func)

    def benchmark_to_format_two_column_only(self):
        """Benchmark each `to_format` method for two-sided column-only
        contexts."""
        for name in ("dict", "records", "df", "arrow"):
            for pivot in PerspectiveBenchmark.COLUMN_PIVOT_OPTIONS:
                if len(pivot) == 0:
                    continue
                test_meta = make_meta(
                    "to_format", "{0}_{1}_column".format(name, len(pivot)))
                view = self._table.view(column_pivots=pivot)
                func = Benchmark(
                    lambda: getattr(view, "to_{0}".format(name))(), meta=test_meta)
                setattr(self, "to_format_{0}".format(test_meta["name"]), func)


if __name__ == "__main__":
    # Initialize a suite and runner, then call `.run()`
    suite = PerspectiveBenchmark()
    runner = Runner(suite)
    runner.run()
